### PR TITLE
705 refactor handlechattogpt to remove mutation

### PR DESCRIPTION
**Pull request by heatherlogan-scottlogic**
## Description

Refactor of  functions inside chatController.ts and openAI.ts to reduce mutation. 
ChatResponse, chatHistroy, sentEmails, should not be mutated across different functions - instead changes should be returned by the using functions either as 

## Notes

<u>openai.ts</u>

* chatGptCallFunction() 
* * <b>Before</b> - mutates sentEmails, declares then mutates reply: ChatCompletionMessageParam
* * <b>Changed</b> - Remove init declaration for reply and construct at the end. 
Split up function calls into handleSendEmailFunction and handleAskQuestionFunction. 
Create a copy of sentEmails and updates this. 
Returns chat completion (with tool call ids), wonLevel and updated sentEmails obj
* performToolCalls()
* * <b>Before</b> - mutates chatResponse (sets wonLevel and defenceReport), calls pushCompletionToHistory()
* * <b>Changed</b> - Removed   chatResponse from input as unused.    Return wonLevel   and defenceReport instead of setting chatResponse parameters.    Clones   history and updates tht    If no toolCalls   then returns original chatResponse.

* getFinalReplyAfterToolCalls()
* * <b>Before</b> - Created blank chatRespond and mutated, calls pushCompletionToHistory and performToolCalls (which previously mutated wonLevel/sentEmails/chatHistory)
* * <b>Changed</b> - Removed blank chatResponse at start (as not needed in chatGptChatCompletion). 
Clones to create sentEmails and updatedChatHistory.
Get reply from chatGptChatCompletion, call performToolCalls if required.
Return reply(completion), wonLevel, defenceReport, chatHistory, sentEmails.

* pushCompletionToHistory()
* * <b>Before</b> - mutates chatHisttory
* * <b>Changed</b> - Clones chatHistory and returns updated version
* chatGptChatCompletion()
* * <b>Before</b> - mutates chatHistory and chatResponse (sets openAIErrorMessage)
* * <b>Changed</b> - eemove chatResponse input parameter. 
Clone chatHistory to apply changes to that depending on system role. 
Returns the completion, updated chat history and error message (possibly null) 

* chatGptSendMessage()
* * <b>Before</b> - mutates chatHistory and chatResponse (sets openAIErrorMessage)
* * <b>Changed</b> - Remove initial blank declaration of ChatResponse that will be mutated
Clones chatHistory and edits that. 
Calls getFinalReplyAfterAllToolCalls which returns new chatResponse, chatHistory, sentEmails

<u>chatController.ts</u>

* handleLowLevelChat()
* * <b>Before</b> - mutates chatResponse
* * <b>Changed</b> - instead returns chatResponse, chatHistory, sentEmails that have been copied/updated by funcs. in openai.ts
* handleHigherLevelChat()
* * <b>Before</b> - mutates chatReponse (lots), mutates req.session.levelState[currentLevel].chatHistory in several places, calls chatGptSendMessage()
* * <b>Changed</b> - Clones the chatReponse and chatHistory objects and applies updates to these, returns chatResponse, chatHistory, sentEmails

* handleChatError()
* * <b>Before</b> - mutates chatResponse, res
* * <b>Changed</b> - Creates updatedChatResponse with blocked/error info and res.sends that. (concern – should we be doing res.send() in chatController? 
* handleErrorGettingReply()
* * <b>Before</b> - handleErrorGettingReply() calls handleChatError(), mutates req (adds error to history)
* * <b>Changed</b> - Change to addErrorToChatHistory to prevent req being mutated in this and handleChatError. Instead returns updated chatHistory, then we can call handleChatError after in the outer function. 
* handleChatToGPT()
* * <b>Before</b> - mutates chatResponse
* * <b>Changed</b> - creates an initial blank chat repsonse obj to use in error handing. then creates an updated obj with updated values from the handleXlevelChat function. updates the state in res.levelState for chatHistory and sentEmails with the final updated values (defences in levelState shouldn't be altered by handleChatToGPT(). sends the updated response 

## Concerns

- Removed the filterOutputDefence as thought it would conflict more with https://github.com/ScottLogic/prompt-injection/pull/740 - TBC. (but this defence doesn't work in this branch currently.. )

## Checklist

Have you done the following?

- [x] Linked the relevant Issue
- [ ] Added tests
- [x] Ensured the workflow steps are passing

### CODE REVIEW

File: backend/test/integration/openai.test.ts [hunk 1]
```diff
@@ -366,134 +375,4 @@ describe('OpenAI Integration Tests', () => {
 			mockCreateChatCompletion.mockRestore();
 		}
 	);
-
-	test('GIVEN the output filtering defence is active WHEN the bot responds with a message containing a phrase in the list THEN the defence is triggered and the message is blocked', async () => {
```
Comment (chriswilty) [inline backend/test/integration/openai.test.ts:370]: Remind me, why did you remove these tests? Are they re-added in a different PR?

**Review by chriswilty — CHANGES_REQUESTED** [review]
Thorough work, good job 👍 

One or two concerns about possible changed behaviour, but i think we can streamline the return values a bit, as discussed on Teams.

**Comment by chriswilty** [comment]
One thing that goes hand-in-hand with immutability is avoiding `let` declarations wherever possible. Sometimes it doesn't seem possible, if we have conditional statements that might or might not overwrite that value, but in that case, it is usually a sign that the function needs to be split. In any case, this is a strong start, and the rest we can tackle as we go.

File: backend/src/controller/handleError.ts [hunk 2]
```diff
@@ -19,14 +18,19 @@ function handleChatError(
 	statusCode = 500
```
[the remark this answers is not in the document]
Comment (heatherlogan-scottlogic) [reply backend/src/controller/handleError.ts:18]: Hmm true i think at one point we did use distinguish between blocked messages and error messages for styling, but i don't think we need this defence information here now.

File: backend/test/integration/openai.test.ts (hunk 1 above)
Comment (heatherlogan-scottlogic) [reply backend/test/integration/openai.test.ts:370]: Yes I had misunderstood and thought George and I would merge branches. Waiting on George confirming after resolving conflicts before adding back in/or not

**Comment by pmarsh-scottlogic** [comment]
Really appreciate the thorough PR description!

**Review by pmarsh-scottlogic — CHANGES_REQUESTED** [review]
Cracking job, love to see it. Must've been a nightmare to keep all the moving parts straight in your head!

Some comments from me, all minor

File: backend/src/openai.ts [hunk 3]
```diff
@@ -333,108 +374,64 @@ function pushCompletionToHistory(
...
-
-			// add the function call to the chat history
-			pushCompletionToHistory(
-				chatHistory,
-				functionCallReply.completion,
-				CHAT_MESSAGE_TYPE.FUNCTION_CALL
-			);
-			// update the defence info
-			chatResponse.defenceReport = functionCallReply.defenceReport;
+			return {
```
[the remark this answers is not in the document]
Comment (heatherlogan-scottlogic) [reply backend/src/openai.ts:421]: For now the only supported tool call is function calls, so that would be the case is the case until openai extends this. I can add a comment to make this clear

File: backend/src/openai.ts [hunk 4]
```diff
@@ -509,40 +515,46 @@ async function chatGptSendMessage(
...
 
-	chatResponse.completion = reply;
+	const chatResponse: ChatResponse = {
+		completion: finalToolCallResponse.gptReply.completion,
+		defenceReport: {
+			blockedReason: '',
+			isBlocked: false,
+			alertedDefences: [],
+			triggeredDefences: [],
+		},
```
[the remark this answers is not in the document]
Comment (heatherlogan-scottlogic) [reply backend/src/openai.ts:533]: good point! I'll look into that now

**Review by chriswilty — COMMENTED** [review]
Nothing else from me. Aside from Peter's open questions (and the merge conflict) this looks good to go.

@pmarsh-scottlogic Would you like to take this one, and I'll handle George's PR?

**Comment by pmarsh-scottlogic** [comment]
> Nothing else from me. Aside from Peter's open questions (and the merge conflict) this looks good to go.
> 
> @pmarsh-scottlogic Would you like to take this one, and I'll handle George's PR?

@chriswilty I was about to suggest exactly the same thing :)

**Review by pmarsh-scottlogic — CHANGES_REQUESTED** [review]
just a few more notes, which I'll take on

File: backend/src/openai.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (pmarsh-scottlogic) [reply backend/src/openai.ts:419]: I've written an investigation ticket #764

File: backend/test/integration/openai.test.ts (hunk 1 above)
Comment (pmarsh-scottlogic) [reply backend/test/integration/openai.test.ts:372]: the tests are being added back in in George's branch [here](https://github.com/ScottLogic/prompt-injection/pull/740/files#r1452379574)

File: backend/src/controller/chatController.ts [hunk 5]
```diff
@@ -129,30 +129,12 @@ async function handleHigherLevelChat(
...
 			wonLevel: openAiReply.chatResponse.wonLevel,
 			openAIErrorMessage: openAiReply.chatResponse.openAIErrorMessage,
-			defenceReport: {
-				...defenceReport,
-				alertedDefences: combinedAlertedDefences,
-				triggeredDefences: combinedTriggeredDefences,
-				isBlocked: openAiReply.chatResponse.defenceReport.isBlocked,
-				blockedReason: openAiReply.chatResponse.defenceReport.blockedReason,
-			},
+			defenceReport,
```
[the remark this answers is not in the document]
Comment (pmarsh-scottlogic) [reply backend/src/controller/chatController.ts:137]: Yup, George's branch adds it back in [here](https://github.com/ScottLogic/prompt-injection/pull/740/files#r1452383741)

**Comment by pmarsh-scottlogic** [comment]
# Review notes for self

- combineChatDefenceReports added to chatController. Did it exist before and is it doing the same thing?
- createNewUserMessages added to chatController. Did it exist before and is it doing the same thing?

## What happens when user sends a message on a low level, say level 1?

- take original history and user message
- add user message to history (type depends on whether it was transformed)
- getFinalReply
  - mutate history to set/unset system role
  - get reply
  - while tool calls
    - mutate history to add reply with type function call
    - mutate chatResponse with new wonLevel value and new defenceReport value
    - if email function call
      - generate sent emails and mutate the emails session variable to add new emails
    - mutate history to add function call completion
- if problems [no reply content or openAiErrorMessage]
  - don't mutate chatResponse further
  - add error message to history
  - send error response
- if no problems
  - run output defences
  - add bot response to history (type depends on whether it was blocked by output defence)
  - mutate chatreponse reply, wonLevel and openAIErrorMessage
  - if blocked
    - push block message to history
    - update sentEmails in response and send httpResponse
  - else (not blocked)
    - update sentEmails in response and send httpResponse

What happens when user sends a message on a high level, say sandbox?

File: backend/src/openai.ts [hunk 6]
```diff
@@ -329,220 +369,127 @@ function getChatCompletionsFromHistory(
...
+		openAIErrorMessage: finalToolCallResponse.gptReply.openAIErrorMessage,
+	};
 
-	if (
-		currentLevel === LEVEL_NAMES.LEVEL_3 ||
-		currentLevel === LEVEL_NAMES.SANDBOX
-	) {
-		applyOutputFilterDefence(reply.content, defences, chatResponse);
+	if (!chatResponse.completion?.content || chatResponse.openAIErrorMessage) {
+		return { chatResponse, chatHistory, sentEmails };
```
Comment (pmarsh-scottlogic) [inline backend/src/openai.ts:485]: Here is a regression. Though is it problematic?
Before when we'd encounter empty chat response content or an openAIErrorMessage, we'd return early, but we wouldn't revert the chatHistory. Now we return the original chatHistory, therefore not including any messages added during `getFinalReplyAfterAllToolCalls`

File: backend/src/controller/chatController.ts [hunk 7]
```diff
@@ -160,90 +220,119 @@ async function handleChatToGPT(req: OpenAiChatRequest, res: Response) {
...
+				role: 'assistant',
+				content: updatedChatResponse.reply,
+			},
+			chatMessageType: CHAT_MESSAGE_TYPE.BOT,
+		});
 	}
 
-	// update sent emails
-	chatResponse.sentEmails =
-		req.session.levelState[currentLevel].sentEmails.slice(numSentEmails);
```
Comment (pmarsh-scottlogic) [inline backend/src/controller/chatController.ts:243]: What the heck is this line meant to do. And do we still want to be doing that?

File: backend/src/controller/chatController.ts [hunk 8]
```diff
@@ -2,136 +2,199 @@ import { Response } from 'express';
...
-		// combine blocked
-		chatResponse.defenceReport.isBlocked = openAiReply.defenceReport.isBlocked;
+	// if blocked, restore original chat history and add user message to chat history without completion
+	const updatedChatHistory = combinedDefenceReport.isBlocked
+		? pushMessageToHistory(chatHistory, {
+				completion: null,
+				chatMessageType: CHAT_MESSAGE_TYPE.USER,
+				infoMessage: message,
+		  })
+		: openAiReply.chatHistory;
```
Comment (pmarsh-scottlogic) [inline backend/src/controller/chatController.ts:174]: I smelled a regression here and did an investigation:

regression:
Suppose the user sends a message in a higher level and the input is blocked, but also when we call getFinalReplyAfterAllToolCalls to get chatGPTReply something goes wrong, and we have an empty response.
Before:
[problem caught] return early out of chatGPTSendMessage, but pass on values for completion, defenceReport, wonLevel and openAIErrorMessage
[input blocked] restore original chat history and mutate history to add user message as info. Slice emails, send response
else set chatReponse defenceReport according to output defences
[output blocked (won't happen because output is "")] (we don't restore history!) mustate history to add block message. Slice emails, send response.
[problem caught] add error message to history and send error response
Now:
continue with new sent emails, wonLevel, openAIErrorMessage, original history
[blocked, but only by input defence since reply is ""] restore original chat history and add user message as infoMessage. Bot reply is set to empty
[problem] mutate chat history to add error, then send error response
update history to add blocked message
mutate state with updated story and sent emails and return response
To conclude:
both before and now, if input message is blocked (by input defences) but also we get an error when getting a reply, the blocking takes precidence over the error and we respond with a "you've been blocked" kind of message ✅
both before and now, if input message not blocked but also we get an error when getting a reply, we won't trigger any output defences, and so we add an error message to history and return an error response ✅
if input is not blocked, we successfully get a reply and the output is blocked, before we would add the block message to the already-mutated history. Now we restore the original history, then add user message as info and then add the block message ❌

File: backend/src/openai.ts (hunk 6 above)
Comment (pmarsh-scottlogic) [reply backend/src/openai.ts:485]: To reason about this let's ask: Why is it important to include the function call in the chat history? Does it change the behaviour of the program?

Yes. If we have sent an email, then if it's in the history then the bot has the context that an email has sent, which might alter a future response. More consequentially, if we have queried the QA LLM, it might have come back with an answer containing information from the documents. If that is kept in the history, then that knowledge is now available to the bot through memory, rather than having to query again.

Now let's consider what the user expects. Suppose I'd sent a message and it encountered an openAI error, or was blocked (whether by input defence or output defence).
- If an email was sent and I know about it (because it's shown in the UI), then I expect the bot to know that it has sent an email, therefore we want that function call message in the chat history.
- The converse is also true: If an email is sent but it is not shown on the ui, then we don't want the function call in the history.
- Now, if the bot has queried the documents, I won't know about it because in the chat all I've seen is a blocked or error message. Therefore probably I don't want the function call in the history.

So, bearing the above in mind, what do we have now, and do we like it?
- what happens if we send an email and the message is blocked. Do we see the email in the ui? Is the function call in the history?
- what happens if we send an email and we get an openAI error. Do we see the email in the ui? Is the function call in the history?
- what happens if we query the document and we get blocked. Is the function call in the history?
- what happens if we query the document and we get an openAI error. Is the function call in the history?

All that being said, the impact on me as the user is minimal. The worse I'll be is slightly confused: "Oh, how did it know about that thing from the document?", "Oh, it thinks it send an email, but I don't think it did. Silly bot.".

File: backend/src/controller/chatController.ts (hunk 8 above)
Comment (pmarsh-scottlogic) [reply backend/src/controller/chatController.ts:174]: see discussion [here](https://github.com/ScottLogic/prompt-injection/pull/743#discussion_r1458894955)

File: backend/src/controller/chatController.ts (hunk 7 above)
Comment (pmarsh-scottlogic) [reply backend/src/controller/chatController.ts:243]: Ah okay, array.slice(index) will return a sub-array from the index, to the end of the array. In practise, this line extracts only those new emails that have been sent as a result of this particular chat message so that they can be returned without returning the entire list of emails. We are still doing this (and now it's much easier to read!), so no regression. ✅

**Review by chriswilty — CHANGES_REQUESTED** [review]
Like you @pmarsh-scottlogic, I have a couple of concerns, plus a handful of very minor things.

File: backend/src/openai.ts (hunk 6 above)
Comment (pmarsh-scottlogic) [reply backend/src/openai.ts:485]: Hmm. after testing I'd argue neither of the behaviour is great.

## Dev branch: error thrown when sending an email
image

## Dev branch: error thrown when sending a winning email
image

## This branch: error thrown when sending an email
image

```json
chatHistory:  [
  {
    completion: {
      role: 'system',
      content: 'Your role is to assist the user with work-related tasks.\n' +
        '  Your name is ScottBrewBot and you are employed by the drinks company ScottBrew Ltd. \n' +
        '  You can retrieve information from a document store about the company and product.\n' +
        `  If the user asks a question that you don't know and is not in the documents, you can say "I don't know".\n` +
        '  You are able to send emails.\n' +
        '  Before sending an email, show the user a draft of the email (including recipient, subject, and body), and ask 
        '  Do not send the email until the user has explicitly confirmed that the draft is correct.'
    },
    chatMessageType: 8
  },
  {
    completion: {
      role: 'user',
      content: `email bob@example.com with subject "What's up" and body "How's it going, dawg?"`
    },
    chatMessageType: 3
  },
  {
    completion: { role: 'assistant', content: null, tool_calls: [Array] },
    chatMessageType: 9
  },
  {
    completion: {
      role: 'tool',
      content: 'Email not sent as not confirmed with user',
      tool_call_id: 'call_4bD8Z4vztASBYL1iipaHYyUx'
    },
    chatMessageType: 9
  },
  {
    completion: {
      role: 'assistant',
      content: 'Sure, I can help you with that. Before sending the email, I will show you a draft of the email. Please co
        '\n' +
        'Recipient: bob@example.com\n' +
        "Subject: What's up\n" +
        "Body: How's it going, dawg?\n" +
        '\n' +
        'Please let me know if I should send this email.'
    },
    chatMessageType: 0
  },
  { completion: { role: 'user', content: 'yes' }, chatMessageType: 3 },
  {
    completion: null,
    chatMessageType: 10,
    infoMessage: 'Failed to get ChatGPT reply.'
  },
  {
    completion: { role: 'user', content: 'did you send that email' },
    chatMessageType: 3
  },
  {
    completion: null,
    chatMessageType: 10,
    infoMessage: 'Failed to get ChatGPT reply.'
  },
  {
    completion: { role: 'user', content: 'did you send that email?' },
    chatMessageType: 3
  },
  {
    completion: {
      role: 'assistant',
      content: 'Apologies for the confusion. Let me resend the email for you.',
      tool_calls: [Array]
    },
    chatMessageType: 9
  },
  {
    completion: {
      role: 'tool',
      content: "Email sent to bob@example.com with subject What's up and body How's it going, dawg?",
      tool_call_id: 'call_6ODZO7iXkEWb1k3Y6MJ1wXl7'
    },
    chatMessageType: 9
  },
  {
    completion: {
      role: 'assistant',
    },
    chatMessageType: 0
  }
]
```

File: backend/src/controller/chatController.ts [hunk 9]
```diff
@@ -2,136 +2,199 @@ import { Response } from 'express';
...
+	const updatedChatResponse: ChatHttpResponse = {
+		...chatResponse,
+		reply: openAiReply.chatResponse.completion?.content?.toString() ?? '',
+		wonLevel: openAiReply.chatResponse.wonLevel,
+		openAIErrorMessage: openAiReply.chatResponse.openAIErrorMessage,
+	};
+	return {
+		chatResponse: updatedChatResponse,
+		chatHistory: openAiReply.chatHistory,
+		sentEmails: openAiReply.sentEmails,
```
[the remark this answers is not in the document]
Comment (pmarsh-scottlogic) [reply backend/src/controller/chatController.ts]: good catch

File: backend/src/controller/chatController.ts [hunk 10]
```diff
@@ -160,90 +220,119 @@ async function handleChatToGPT(req: OpenAiChatRequest, res: Response) {
...
 			: defaultChatModel;
 
-	// record the history before chat completion called
-	const chatHistoryBefore = [
+	const currentChatHistory = [
 		...req.session.levelState[currentLevel].chatHistory,
 	];
+	const defences = [...req.session.levelState[currentLevel].defences];
+
+	let levelResult = null;
```
[the remark this answers is not in the document]
Comment (pmarsh-scottlogic) [reply backend/src/controller/chatController.ts:242]: Not allowed by linter! If we were allowed, what would the default value be?

Ideally though we'd have something like

```typescript
const chatHandler = currentLevel < LEVEL_NAMES.LEVEL_3 ?
  handleChatWithoutDefenceDetection :
  handleChatWithDefenceDetection;
const levelResult = chatHandler (...).catch((error) => {
  // handle error
})
``` 

but I'm scared to make more changes, so can we leave that for the future?

File: backend/src/controller/chatController.ts [hunk 11]
```diff
@@ -2,136 +2,199 @@ import { Response } from 'express';
...
+	const combinedReport: ChatDefenceReport = {
+		blockedReason: reports
+			.filter((report) => report.blockedReason !== null)
+			.map((report) => report.blockedReason)
+			.join('\n'),
+		isBlocked: reports.some((report) => report.isBlocked),
+		alertedDefences: reports.flatMap((report) => report.alertedDefences),
+		triggeredDefences: reports.flatMap((report) => report.triggeredDefences),
+	};
+	return combinedReport;
```
[the remark this answers is not in the document]
Comment (pmarsh-scottlogic) [reply backend/src/controller/chatController.ts:42]: does your IDE come with its own linter? I'm not seeing that warning

[the remark this answers is not in the document]
Comment (pmarsh-scottlogic) [reply backend/src/controller/chatController.ts]: change made

File: backend/src/openai.ts [hunk 12]
```diff
@@ -329,220 +369,127 @@ function getChatCompletionsFromHistory(
...
-			chatResponse.wonLevel = functionCallReply.wonLevel;
-
-			// add the function call to the chat history
-			pushCompletionToHistory(
-				chatHistory,
-				functionCallReply.completion,
-				CHAT_MESSAGE_TYPE.FUNCTION_CALL
-			);
-			// update the defence info
-			chatResponse.defenceReport = functionCallReply.defenceReport;
```
[the remark this answers is not in the document]
Comment (pmarsh-scottlogic) [reply backend/src/openai.ts:435]: It's all good!

Originally we were passing `defenceReport` into `chatGptCallFunction`, and doing precisely zero processing on it / with it before spitting it out again in the return object. That is to say, we never needed to pass it in or out in the first place. After all, the only things we'd do in `chatGptCallFunction` is look at docs or send an email, and currently there are no defences that interact with those systems.

Now, all the defence detection calls live inside `handleChatWithDefenceDetection` in _chatController.ts_. Which is nice because `chatGptSendMessage` is now solely responsible for getting a reply from gpt and dealing with function calls. The input defence detection happens before that, and the output defence detection happens after, and all the concerns are happily separated (mostly).

File: backend/src/controller/handleError.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmarsh-scottlogic) [reply backend/src/controller/handleError.ts:17]: I suppose you mean `isError` is always true, when we've sent the response via `handleChatError`.

Anyway, luckily not much tracing to do, all you need to do is look at `processChatResponse` in _ChatBox.tsx_. Yes, when `isError` is true, the `defenceReport `is irrelevant. But the `defenceReport` is relevant if `isError` is false. And in the frontend, we don't know ahead of time whether sending a message will cause an error or not, therefore we still need `isError` and `defenceReport` in `ChatHttpResponse`.

File: backend/src/openai.ts [hunk 13]
```diff
@@ -329,220 +369,127 @@ function getChatCompletionsFromHistory(
...
+				defences,
+				currentLevel
+			);
+
+			updatedChatHistory = toolCallReply.chatHistory;
+			if (toolCallReply.functionCallReply?.sentEmails) {
+				sentEmails.push(...toolCallReply.functionCallReply.sentEmails);
+			}
+			wonLevel =
+				(wonLevel || toolCallReply.functionCallReply?.wonLevel) ?? false;
```
[the remark this answers is not in the document]
Comment (pmarsh-scottlogic) [reply backend/src/openai.ts:447]: see [my reply to your other comment](https://github.com/ScottLogic/prompt-injection/pull/743#discussion_r1459247664)

File: backend/src/openai.ts (hunk 6 above)
Comment (pmarsh-scottlogic) [reply backend/src/openai.ts:484]: my vote is that if we encounter an error while confirming an email, we don't send the email. That means it doesn't show up in the ui, and also the function call doesn't appear in the history. The processing would _so much simpler_ that way.

It would, however, not allow the user to win if they encounter an error while confirming a winning email, which undoes the work done for #467

File: backend/src/controller/handleError.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmarsh-scottlogic) [reply backend/src/controller/handleError.ts:17]: I'm just checking the removed code now

[the remark this answers is not in the document]
Comment (chriswilty) [reply backend/src/controller/handleError.ts:17]: Does this mean we _could_ send null / undefined for defenceReport when we have a chat error? If so, then `defenceReport` in `ChatHttpResponse` would need to be optional, which is maybe not great for type-safety. In that case, maybe we could split the response into two types: ChatHttpResponse and ChatHttpErrorResponse.

That might be one to think about for future improvements though!

File: backend/src/openai.ts (hunk 6 above)
Comment (chriswilty) [reply backend/src/openai.ts:484]: Yep that seems to make sense to me. The user just needs to try again.

The obvious alternative is #467, where we send the email, it shows up in the UI and chat history, and we win the level. That is a tad confusing for the user cos they've seen an error message. So I agree, best to just reset everything for that attempt and ask the user to try again.

File: backend/src/controller/chatController.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (chriswilty) [reply backend/src/controller/chatController.ts:241]: Why does the linter complain, what's the rule violation? Let me check...

Ah, it's a rule we are explicitly turning on in our linter config:

```
'@typescript-eslint/init-declarations': 'error',
```

I wonder why. It's fairly common to set the value of a `let` conditionally, particularly when the logic is a little complex. Using ternary assignment is a possibility, but it can become a bit gnarly, particularly if there are try-catches involved. The compiler will complain anyway if a `let` defined in this way is used when it may not have been initialised, and a `const` must be declared with a value anyway.

I'd vote for removing this rule from our linter config. Frontend also, for consistency. Note it's not a rule that's included in any of the recommended rulesets we are using, so we can just remove the lines, instead of turning them off.

File: backend/src/controller/chatController.ts (hunk 7 above)
Comment (chriswilty) [reply backend/src/controller/chatController.ts:243]: Haha, you explained it so much better than me 😅

File: backend/src/openai.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (chriswilty) [reply backend/src/openai.ts:435]: Thank you, excellent explanation! You can tell I wrote this comment on a friday afternoon of a hectic week, cos I missed that all-important detail that we weren't even doing anything with it in the original call 😅

**Comment by chriswilty** [comment]
@pmarsh-scottlogic I'm very happy with this now, good work. The linter rule is small beans, remove it if you wish - doesn't stop me approving.

**Review by chriswilty — APPROVED** [review]
👌 Awesome work!

File: backend/src/controller/handleError.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmarsh-scottlogic) [reply backend/src/controller/handleError.ts:17]: the `defenceReport` in the response is necessarily not null/undefined! It just might be "empty", that is it might have a null `blockedReason`, or an empty `triggeredDefences` list. Which means the response is bigger than it needs to be when defences are not active, but It's fine for now

File: backend/src/openai.ts (hunk 6 above)
Comment (pmarsh-scottlogic) [reply backend/src/openai.ts:484]: Lovely. I've made ticket #797  capture that
